warp10/src/main/java/io/warp10/script/ext/http/HTTP.java

```diff
@@ -0,0 +1,533 @@
+      urlCount = new AtomicLong();
+      downloadSize = new AtomicLong();
+      stack.setAttribute(HttpWarpScriptExtension.ATTRIBUTE_HTTP_REQUESTS, urlCount);
+      stack.setAttribute(HttpWarpScriptExtension.ATTRIBUTE_HTTP_SIZE, downloadSize);
+    } else {
+      urlCount = (AtomicLong) ufCount;
+      downloadSize = (AtomicLong) ufSize;
+    }
+
+    if (urlCount.addAndGet(1) > maxrequests) {
+      throw new WarpScriptException(getName() + " is limited to " + maxrequests + " calls per script execution.");
+    }
+
+    Map<String, Object> res = new LinkedHashMap<>();
+    HttpURLConnection conn = null;
+
+    try {
+      conn = (HttpURLConnection) url.openConnection();
+
+      //
+      // Set headers
+      //
+
+      if (null == headersMacro) {
+
+        Object username = params.get(USERNAME);
+        Object password = params.get(PASSWORD);
+
+        if (null != username && null != password) {
+
+          //
+          // Compute basic auth
+          //
+
+          if (!(username instanceof String)) {
+            throw new WarpScriptException(getName() + " expects a STRING username when doing basic authentication.");
+          }
+
+          if (!(password instanceof String)) {
+            throw new WarpScriptException(getName() + " expects a STRING password when doing basic authentication.");
+          }
+
+          String userInfo = ((String) username) + ":" + ((String) password);
+          String basicAuth = "Basic " + Base64.encodeBase64String(userInfo.getBytes(StandardCharsets.UTF_8));
+          conn.setRequestProperty("Authorization", basicAuth);
+        }
+
+      } else {
+
+        stack.push(params);
+        stack.exec(headersMacro);
+        o = stack.pop();
+
+        if (!(o instanceof Map)) {
+          throw new WarpScriptException(getName() + " expects the " + HEADERS_MACRO + " to push a headers map onto the stack.");
+        }
+
+        headers = (Map) o;
+      }
+
+      for (Map.Entry<Object, Object> prop: headers.entrySet()) {
+        conn.setRequestProperty(String.valueOf(prop.getKey()), String.valueOf(prop.getValue()));
+      }
+
+      conn.setDoInput(true);
+      conn.setRequestMethod(method.toUpperCase());
+
+      //
+      // Set body
+      //
+
+      if ("GET" != method && "DELETE" != method && "TRACE" != method && "OPTIONS" != method && "HEAD" != method) {
+
+        if (body instanceof String) {
+          String bodyS = (String) body;
+          conn.setDoOutput(bodyS.length() > 0);
+          if (bodyS.length() > 0) {
+            try (OutputStream os = conn.getOutputStream()) {
+              os.write(bodyS.getBytes(StandardCharsets.UTF_8));
+            }
+          }
+
+        } else if (body instanceof byte[]) {
+          byte[] bodyB = (byte[]) body;
+          conn.setDoOutput(bodyB.length > 0);
+          if (bodyB.length > 0) {
+            try (OutputStream os = conn.getOutputStream()) {
+              os.write(bodyB);
+            }
+          }
```

Comment:
Don't duplicate the code, simply convert the String to byte[], with `bodyS.getBytes(StandardCharsets.UTF_8)`, then do the logic on the body as byte[].